python/pyarrow/tests/test_extension_type.py

```diff
@@ -217,3 +217,50 @@ def test_ipc_unknown_type():
     batch = ipc_read_batch(buf2)
     arr = check_example_batch(batch)
     assert arr.type == ParamExtType(3)
+
+
+class PeriodType(pa.GenericExtensionType):
```

Comment:
this is an example of how the API would look like for people implementing an extension type in Python.

Here, as initial example, I took the "period" type from pandas (which is currently not supported by pyarrow out of the box), which is basically integer values with a frequency.

Reply:
Unfortunately I'm having trouble understanding the intended interface from this example as it's too simple - AFAICS it effectively boils down to:
```python
    def __arrow_ext_serialize__(self):
        return self.freq.encode()

    @classmethod
    def __arrow_ext_deserialize__(cls, storage_type, serialized):
        freq = serialized.decode()
        return PeriodType(freq)
```
i.e. it's really just passing a python string and then using the same string to instantiate the class - no serialization or other involvement with arrow that I can see.

I assume the api of the serialize method is is:
```python
def __arrow_ext_serialize__(self) -> bytes:
    ...
```
(documentation using type hints to show the api would be extremely helpful - especially since the actually implementation will be hidden in a compiled module)

I can decompose my python types into (possibly nested) dictionaries mapping names to simple types where simple types include stings, numbers, timestamps or arrays/dsataframes of those types. How do I then turn that into a single bytearray to return from `__arrow_ext_serialize__`?

Reply:
> i.e. it's really just passing a python string and then using the same string to instantiate the class - no serialization or other involvement with arrow that I can see.

Yes, that is correct. But what aspect do you not understand about this example? (It's simple, but still an actual real world use case (that I intend to implement in pandas), so that it can be made simple seems like a plus)

> I can decompose my python types into (possibly nested) dictionaries mapping names to simple types where simple types include stings, numbers, timestamps or arrays/dsataframes of those types. How do I then turn that into a single bytearray to return from __arrow_ext_serialize__?

Can you give a more concrete example of what you are trying to achieve here?

Reply:
I understand the example fine - I just don't know how to use it to implement anything not trivial.

The example is trivial because it's not using arrow at all. It serializes a python string by calling the `encode` method and `deserializes` by calling `decode`. This works for strings but doesn't say anything about how to convert any other python object (which doesn't have handy `encode` /`decode` method)  into bytes.